File: src/commands/reconciliation.ts
import {
  ReconciliationSession,
  ReconciliationItem,
  ReconciliationDiscrepancy,
  ReconciliationSummary,
  CreateReconciliationSessionRequest,
  AddReconciliationItemRequest,
} from "@/lib/types";
import { getRunEnv, RUN_ENV, invokeTauri, invokeWeb, logger } from "@/adapters";

// ============================================================================
// Session Operations
// ============================================================================

export const createReconciliationSession = async (
  request: CreateReconciliationSessionRequest,
): Promise<ReconciliationSession> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("create_reconciliation_session", { request });
      case RUN_ENV.WEB:
        return invokeWeb("create_reconciliation_session", { request });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error creating reconciliation session.");
    throw error;
  }
};

export const getReconciliationSession = async (
  sessionId: string,
): Promise<ReconciliationSession> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_reconciliation_session", { sessionId });
      case RUN_ENV.WEB:
        return invokeWeb("get_reconciliation_session", { sessionId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching reconciliation session.");
    throw error;
  }
};

export const getReconciliationSessionsByAccount = async (
  accountId: string,
): Promise<ReconciliationSession[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_reconciliation_sessions_by_account", { accountId });
      case RUN_ENV.WEB:
        return invokeWeb("get_reconciliation_sessions_by_account", { accountId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching reconciliation sessions.");
    throw error;
  }
};

export const getReconciliationSummary = async (
  sessionId: string,
): Promise<ReconciliationSummary> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_reconciliation_summary", { sessionId });
      case RUN_ENV.WEB:
        return invokeWeb("get_reconciliation_summary", { sessionId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching reconciliation summary.");
    throw error;
  }
};

export const completeReconciliationSession = async (
  sessionId: string,
): Promise<ReconciliationSession> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("complete_reconciliation_session", { sessionId });
      case RUN_ENV.WEB:
        return invokeWeb("complete_reconciliation_session", { sessionId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error completing reconciliation session.");
    throw error;
  }
};

export const cancelReconciliationSession = async (
  sessionId: string,
): Promise<ReconciliationSession> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("cancel_reconciliation_session", { sessionId });
      case RUN_ENV.WEB:
        return invokeWeb("cancel_reconciliation_session", { sessionId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error cancelling reconciliation session.");
    throw error;
  }
};

export const deleteReconciliationSession = async (sessionId: string): Promise<void> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("delete_reconciliation_session", { sessionId });
      case RUN_ENV.WEB:
        return invokeWeb("delete_reconciliation_session", { sessionId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error deleting reconciliation session.");
    throw error;
  }
};

// ============================================================================
// Item Operations (Broker Statement Entries)
// ============================================================================

export const addReconciliationItem = async (
  request: AddReconciliationItemRequest,
): Promise<ReconciliationItem> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("add_reconciliation_item", { request });
      case RUN_ENV.WEB:
        return invokeWeb("add_reconciliation_item", { request });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error adding reconciliation item.");
    throw error;
  }
};

export const addReconciliationItems = async (
  requests: AddReconciliationItemRequest[],
): Promise<ReconciliationItem[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("add_reconciliation_items", { requests });
      case RUN_ENV.WEB:
        return invokeWeb("add_reconciliation_items", { requests });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error adding reconciliation items.");
    throw error;
  }
};

export const getReconciliationItems = async (
  sessionId: string,
): Promise<ReconciliationItem[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_reconciliation_items", { sessionId });
      case RUN_ENV.WEB:
        return invokeWeb("get_reconciliation_items", { sessionId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching reconciliation items.");
    throw error;
  }
};

export const deleteReconciliationItem = async (itemId: string): Promise<void> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("delete_reconciliation_item", { itemId });
      case RUN_ENV.WEB:
        return invokeWeb("delete_reconciliation_item", { itemId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error deleting reconciliation item.");
    throw error;
  }
};

// ============================================================================
// Discrepancy Operations
// ============================================================================

export const calculateDiscrepancies = async (
  sessionId: string,
): Promise<ReconciliationDiscrepancy[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("calculate_discrepancies", { sessionId });
      case RUN_ENV.WEB:
        return invokeWeb("calculate_discrepancies", { sessionId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error calculating discrepancies.");
    throw error;
  }
};

export const getDiscrepancies = async (
  sessionId: string,
): Promise<ReconciliationDiscrepancy[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_discrepancies", { sessionId });
      case RUN_ENV.WEB:
        return invokeWeb("get_discrepancies", { sessionId });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error fetching discrepancies.");
    throw error;
  }
};

export const resolveDiscrepancy = async (
  discrepancyId: string,
  resolutionActivityId?: string,
  notes?: string,
): Promise<ReconciliationDiscrepancy> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("resolve_discrepancy", {
          discrepancyId,
          resolutionActivityId,
          notes,
        });
      case RUN_ENV.WEB:
        return invokeWeb("resolve_discrepancy", {
          discrepancyId,
          resolutionActivityId,
          notes,
        });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error resolving discrepancy.");
    throw error;
  }
};

export const ignoreDiscrepancy = async (
  discrepancyId: string,
  notes?: string,
): Promise<ReconciliationDiscrepancy> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("ignore_discrepancy", { discrepancyId, notes });
      case RUN_ENV.WEB:
        return invokeWeb("ignore_discrepancy", { discrepancyId, notes });
      default:
        throw new Error(`Unsupported`);
    }
  } catch (error) {
    logger.error("Error ignoring discrepancy.");
    throw error;
  }
};

File: src-tauri/src/commands/reconciliation.rs
use std::sync::Arc;

use crate::context::ServiceContext;
use log::debug;
use tauri::State;
use wealthfolio_core::reconciliation::{
    AddReconciliationItemRequest, CreateReconciliationSessionRequest, ReconciliationDiscrepancy,
    ReconciliationItem, ReconciliationSession, ReconciliationSummary,
};

// ============================================================================
// Session Commands
// ============================================================================

#[tauri::command]
pub async fn create_reconciliation_session(
    request: CreateReconciliationSessionRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ReconciliationSession, String> {
    debug!("Creating reconciliation session...");
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::create_session(&mut conn, request)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_reconciliation_session(
    session_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ReconciliationSession, String> {
    debug!("Getting reconciliation session: {}", session_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::get_session(&mut conn, &session_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_reconciliation_sessions_by_account(
    account_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ReconciliationSession>, String> {
    debug!("Getting reconciliation sessions for account: {}", account_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::get_sessions_by_account(
        &mut conn,
        &account_id,
    )
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_reconciliation_summary(
    session_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ReconciliationSummary, String> {
    debug!("Getting reconciliation summary: {}", session_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::get_session_summary(
        &mut conn,
        &session_id,
    )
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn complete_reconciliation_session(
    session_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ReconciliationSession, String> {
    debug!("Completing reconciliation session: {}", session_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::complete_session(&mut conn, &session_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn cancel_reconciliation_session(
    session_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ReconciliationSession, String> {
    debug!("Cancelling reconciliation session: {}", session_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::cancel_session(&mut conn, &session_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_reconciliation_session(
    session_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<(), String> {
    debug!("Deleting reconciliation session: {}", session_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::delete_session(&mut conn, &session_id)
        .map_err(|e| e.to_string())
}

// ============================================================================
// Item Commands (Broker Statement Entries)
// ============================================================================

#[tauri::command]
pub async fn add_reconciliation_item(
    request: AddReconciliationItemRequest,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ReconciliationItem, String> {
    debug!("Adding reconciliation item...");
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::add_item(&mut conn, request)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn add_reconciliation_items(
    requests: Vec<AddReconciliationItemRequest>,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ReconciliationItem>, String> {
    debug!("Adding {} reconciliation items...", requests.len());
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::add_items(&mut conn, requests)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_reconciliation_items(
    session_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ReconciliationItem>, String> {
    debug!("Getting reconciliation items for session: {}", session_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::get_items_by_session(
        &mut conn,
        &session_id,
    )
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_reconciliation_item(
    item_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<(), String> {
    debug!("Deleting reconciliation item: {}", item_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::delete_item(&mut conn, &item_id)
        .map_err(|e| e.to_string())
}

// ============================================================================
// Discrepancy Commands
// ============================================================================

#[tauri::command]
pub async fn calculate_discrepancies(
    session_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ReconciliationDiscrepancy>, String> {
    debug!("Calculating discrepancies for session: {}", session_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::calculate_discrepancies(
        &mut conn,
        &session_id,
    )
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_discrepancies(
    session_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<ReconciliationDiscrepancy>, String> {
    debug!("Getting discrepancies for session: {}", session_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::get_discrepancies(&mut conn, &session_id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn resolve_discrepancy(
    discrepancy_id: String,
    resolution_activity_id: Option<String>,
    notes: Option<String>,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ReconciliationDiscrepancy, String> {
    debug!("Resolving discrepancy: {}", discrepancy_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::resolve_discrepancy(
        &mut conn,
        &discrepancy_id,
        resolution_activity_id,
        notes,
    )
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn ignore_discrepancy(
    discrepancy_id: String,
    notes: Option<String>,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<ReconciliationDiscrepancy, String> {
    debug!("Ignoring discrepancy: {}", discrepancy_id);
    let mut conn = state.pool().get().map_err(|e| e.to_string())?;
    wealthfolio_core::reconciliation::ReconciliationService::ignore_discrepancy(
        &mut conn,
        &discrepancy_id,
        notes,
    )
    .map_err(|e| e.to_string())
}
